### PR TITLE
Add Conv3DLayer, DeConv3DLayer

### PR DESCRIPTION
**Pull request by chengduoZH**
fix https://github.com/PaddlePaddle/Paddle/issues/3359
Add Conv3DLayer, DeConv3DLayer.

* [x] CPU implementation. 
* [x] GPU implementation.
* [x] Add test_Conv3D and Test_DeConv3D unit test.
* [x] Pass CPU unit test
* [x] Pass GPU unit test
* [x] Add config parser for Conv3D and DeConv3D

### CODE REVIEW

File: paddle/gserver/layers/Conv3DLayer.cpp [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+                           paddingZ_[i], paddingY_[i], padding_[i]);
+
+          real *outData = outMat->getData() + n * outWidth;
+          MatrixPtr outMatSub =
+                  Matrix::create(outData, groups_[i] * M, N, false, useGpu_);
+          for (int g = 0; g < groups_[i]; g++) {
+              MatrixPtr wMatSub = wMat->subMatrix(g * M, M);
+              MatrixPtr in = colBuf_->subMatrix(g * K, K);
+              MatrixPtr out = outMatSub->subMatrix(g * M, M);
+              out->mul(*wMatSub, *in, 1.0, 0.0);
```
Comment (hedaoyuan) [inline paddle/gserver/layers/Conv3DLayer.cpp]: 0.0需要修改成1.0，否则inputLayers_.size() > 1时就错了。

File: paddle/gserver/layers/ConvBaseLayer.cpp [hunk 2]
```diff
@@ -21,9 +21,11 @@ bool ConvBaseLayer::init(const LayerMap& layerMap,
                          const ParameterMap& parameterMap) {
   /* Initialize the basic parent class */
   Layer::init(layerMap, parameterMap);
-  isDeconv_ = (config_.type() == "exconv" || config_.type() == "cudnn_conv")
-                  ? false
-                  : true;
+  isDeconv_ = (config_.type() == "exconv" ||
+          config_.type() == "cudnn_conv" ||
+          config_.type() == "conv3d" ||
+          config_.type() == "deconv3d"   )
```
Comment (hedaoyuan) [inline paddle/gserver/layers/ConvBaseLayer.cpp]: 27行不能有。

File: paddle/gserver/layers/DeConv3DLayer.cpp [hunk 3]
```diff
@@ -0,0 +1,211 @@
...
+          MatrixPtr inMatSub = Matrix::create(
+                  inData, M, N, false, useGpu_);
+          MatrixPtr wGradMatSub = Matrix::create(
+                  wGradData, K, M, false, useGpu_);
+          wGradMatSub->mul(*colBufDataSub,
+                  *(inMatSub->getTranspose()), 1.0, 1.0);
+          wGradData += K * M;
+          inData += M * N;
+        }
+        weights_[i]->getParameterPtr()->incUpdate(callback);
```
Comment (hedaoyuan) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: 171行放到batchSize循环外面。

File: paddle/gserver/layers/ConvBaseLayer.cpp (hunk 2 above)
Comment (chengduoZH) [inline paddle/gserver/layers/ConvBaseLayer.cpp]: 这个得有，不然会出错，在DeConv3D中filter数据排布方式是： (output * depth * height * weigth) x  channel，所以```width = (!isDeconv_) ? numFilters_ : channels_[i];```中width要等与numFilters_

File: paddle/gserver/layers/Conv3DLayer.cpp [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+    if (nullptr != weights_[index]->getW())
+      weights_[index]->getW()->reshape(weights_[index]->getW()->getWidth(),
+                                       weights_[index]->getW()->getHeight());
+    if (nullptr != weights_[index]->getWGrad())
+      weights_[index]->getWGrad()->reshape(
+          weights_[index]->getWGrad()->getWidth(),
+          weights_[index]->getWGrad()->getHeight());
+    ++index;
+  }
+  CHECK(inputLayers_.size() == parameters_.size());
```
Comment (qingqing01) [inline paddle/gserver/layers/Conv3DLayer.cpp]: This line is not correct. Remove this check. If `inputLayers_.size() == 1` and this layer has bias, the `parameters_.size() == 2`, so it is not correct.

File: paddle/gserver/layers/Conv3DLayer.cpp [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+  return true;
+}
+
+size_t Conv3DLayer::getSize() {
+  CHECK_NE(inputLayers_.size(), 0UL);
+  // imgSizeH_.clear();
+  // imgSizeW_.clear();
+  // imgSizeD_.clear();
```
Comment (qingqing01) [inline paddle/gserver/layers/Conv3DLayer.cpp]: Remove the commented code.

File: paddle/gserver/layers/ConvBaseLayer.cpp [hunk 6]
```diff
@@ -21,7 +21,8 @@ bool ConvBaseLayer::init(const LayerMap& layerMap,
                          const ParameterMap& parameterMap) {
   /* Initialize the basic parent class */
   Layer::init(layerMap, parameterMap);
-  isDeconv_ = (config_.type() == "exconv" || config_.type() == "cudnn_conv")
+  isDeconv_ = (config_.type() == "exconv" || config_.type() == "cudnn_conv" ||
+               config_.type() == "conv3d" || config_.type() == "deconv3d")
```
Comment (qingqing01) [inline paddle/gserver/layers/ConvBaseLayer.cpp]: That's not a very good way to specify 2D DeConv.

File: paddle/gserver/layers/DeConv3DLayer.cpp [hunk 7]
```diff
@@ -0,0 +1,210 @@
...
+    M_.push_back(filterChannels_[index]);
+    K_.push_back(filterPixels_[index] * (numFilters_ / groups_[index]));
+    if (weights_[index]->getW())
+      weights_[index]->getW()->reshape(filterPixels_[index] * numFilters_,
+                                       filterChannels_[index]);
+    if (weights_[index]->getWGrad())
+      weights_[index]->getWGrad()->reshape(filterPixels_[index] * numFilters_,
+                                           filterChannels_[index]);
+  }
+  CHECK(inputLayers_.size() == parameters_.size());
```
Comment (qingqing01) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: remove this check.

File: paddle/gserver/layers/DeConv3DLayer.cpp [hunk 8]
```diff
@@ -0,0 +1,210 @@
...
+  // Matrix storage format: (output * depth * height * weigth) x  channel
+  for (int index = 0; index < config_.inputs().size(); ++index) {
+    M_.push_back(filterChannels_[index]);
+    K_.push_back(filterPixels_[index] * (numFilters_ / groups_[index]));
+    if (weights_[index]->getW())
+      weights_[index]->getW()->reshape(filterPixels_[index] * numFilters_,
+                                       filterChannels_[index]);
+    if (weights_[index]->getWGrad())
+      weights_[index]->getWGrad()->reshape(filterPixels_[index] * numFilters_,
+                                           filterChannels_[index]);
```
Comment (qingqing01) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: Same problem as the 3DConv.

File: paddle/gserver/layers/DeConv3DLayer.cpp [hunk 9]
```diff
@@ -0,0 +1,210 @@
...
+size_t DeConv3DLayer::getSize() {
+  CHECK_NE(inputLayers_.size(), 0UL);
+  // imgSizeH_.clear();
+  // imgSizeW_.clear();
+  // imgSizeD_.clear();
+  outputH_.clear();
+  outputW_.clear();
+  outputD_.clear();
+  N_.clear();
+  No_.clear();
```
Comment (qingqing01) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: No_ -> NOut_

File: paddle/gserver/layers/DeConv3DLayer.cpp [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+}
+
+void DeConv3DLayer::addBias() {
+  MatrixPtr outMat = getOutputValue();
+  if (this->sharedBiases_) {
+    outMat->addSharedBias(*(biases_->getW()), 1.0f);
+  } else {
+    outMat->addBias(*(biases_->getW()), 1.0f);
+  }
+}
```
Comment (qingqing01) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: In fact, the same problem as the implementing 2D-Deconv before. Many similar code with 3D-Conv.

File: paddle/gserver/layers/DeConv3DLayer.cpp [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+#include "DeConv3DLayer.h"
+#include "paddle/utils/Logging.h"
+#include "paddle/utils/Stat.h"
+
+namespace paddle {
+
+REGISTER_LAYER(deconv3d, DeConv3DLayer);
+
+#define DECONV_OUTPUT_SIZE(IN_SIZE, STRID, PAD, KSIZE) \
+  (((IN_SIZE)-1) * (STRID)-2 * (PAD) + (KSIZE))
```
Comment (qingqing01) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: Why not the use `imageSize()` in `./paddle/math/MathUtils.h`?

File: paddle/gserver/tests/test_LayerGrad.cpp [hunk 12]
```diff
@@ -2007,6 +2007,158 @@ TEST(Layer, RowL2NormLayer) {
...
+
+TEST(Layer, test3DConvLayer) {
+  test3DConvLayer("conv3d", /* trans= */ false, /* useGpu= */ false);
+#ifndef PADDLE_ONLY_CPU
+  test3DConvLayer("conv3d", /* trans= */ false, /* useGpu= */ true);
+#endif
+}
+
+int deConvOutputSize(int inSize, int kSize, int pad, int stride) {
+  return (inSize - 1) * stride - 2 * pad + kSize;
```
Comment (qingqing01) [inline paddle/gserver/tests/test_LayerGrad.cpp]: Why not the use `imageSize()` in `./paddle/math/MathUtils.h`?

File: paddle/math/tests/test_matrixCompare.cpp [hunk 13]
```diff
@@ -1203,4 +1203,120 @@ TEST(Matrix, warpCTC) {
   }
 }
 
+int outputSizeCol2Vol(
```
Comment (qingqing01) [inline paddle/math/tests/test_matrixCompare.cpp]: same as above comments.  `./paddle/math/MathUtils.h` has contains the calculation.

File: paddle/math/tests/test_matrixCompare.cpp [hunk 14]
```diff
@@ -1203,4 +1203,120 @@ TEST(Matrix, warpCTC) {
...
+                     1.0,
+                     1.0);
+  TensorCheckErr(*cpuImage, *gpuImage);
+}
+
+TEST(Matrix, col2Vol) {
+  for (auto depth : {9, 16, 64, 128}) {
+    for (auto height : {9, 11, 73, 128, 256}) {
+      for (auto width : {
+               9, 32, 100, 512,
```
Comment (qingqing01) [inline paddle/math/tests/test_matrixCompare.cpp]: For speeding testing, please use the smaller number for depth, height, width.  But please test larger number off-line.

File: paddle/gserver/layers/Conv3DLayer.cpp (hunk 4 above)
Comment (chengduoZH) [inline paddle/gserver/layers/Conv3DLayer.cpp]: So the 58 line of ConqbaseLayer.cpp should be removed?

File: paddle/gserver/layers/Conv3DLayer.cpp [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+  int index = 0;
+  for (auto &inputConfig : config_.inputs()) {
+      const ConvConfig &conf = inputConfig.conv_conf();
+      M_.push_back(numFilters_ / conf.groups());
+      K_.push_back(
+              conf.filter_channels() * conf.filter_size_z() * \
+      conf.filter_size_y() * conf.filter_size());
+      weights_[index]->getW()->reshape(
+              weights_[index]->getW()->getWidth(),
+              weights_[index]->getW()->getHeight());
```
Comment (chengduoZH) [inline paddle/gserver/layers/Conv3DLayer.cpp]: 在ConqBaseLayer中创建的weights_的shape在Conv3DLayer中不适用，所以在Conv3DLayer中需要把weights_的width和height颠倒过来，不过这个在最新的提交中做了做了调整，即把创建weights_的操作写在派生类中

File: paddle/gserver/layers/Conv3DLayer.cpp (hunk 1 above)
Comment (chengduoZH) [inline paddle/gserver/layers/Conv3DLayer.cpp]: Done

File: paddle/gserver/layers/DeConv3DLayer.cpp (hunk 3 above)
Comment (chengduoZH) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: Done

File: paddle/math/Matrix.h [hunk 16]
```diff
@@ -1039,6 +1039,42 @@ class Matrix : public BaseMatrix {
     LOG(FATAL) << "Not implemented";
   }
 
+  virtual void vol2Col(real* data,
```
Comment (chengduoZH) [inline paddle/math/Matrix.h]: Done

File: paddle/cuda/src/hl_cuda_matrix.cu [hunk 17]
```diff
@@ -592,3 +592,138 @@ void hl_matrix_rotate(
       mat, matRot, dimM, dimN, clockWise);
   CHECK_SYNC("hl_matrix_rotate failed");
 }
+
+
+__global__ void keMatrixVol2Col(
+        int num_kernels, real*data, real* data_col,
```
Comment (chengduoZH) [inline paddle/cuda/src/hl_cuda_matrix.cu]: Done

File: paddle/cuda/src/hl_cuda_matrix.cu [hunk 18]
```diff
@@ -592,3 +592,204 @@ void hl_matrix_rotate(
...
+                       real* dataDst) {
+  int depth_col = (depth + 2 * paddingD - filterD) / strideD + 1;
+  int height_col = (height + 2 * paddingH - filterH) / strideH + 1;
+  int width_col = (width + 2 * paddingW - filterW) / strideW + 1;
+  int num_kernels = channels * depth_col * height_col * width_col;
+
+  const int threads = 512;
+  const int blocks = DIVUP(num_kernels, threads);
+
+  keMatrixVol2Col<<<blocks, threads>>>(num_kernels,
```
Comment (chengduoZH) [inline paddle/cuda/src/hl_cuda_matrix.cu]: Done

File: paddle/cuda/src/hl_cuda_matrix.cu [hunk 19]
```diff
@@ -592,3 +592,204 @@ void hl_matrix_rotate(
...
+                       real beta) {
+  int depth_col = (depth + 2 * paddingD - filterD) / strideD + 1;
+  int height_col = (height + 2 * paddingH - filterH) / strideH + 1;
+  int width_col = (width + 2 * paddingW - filterW) / strideW + 1;
+  int num_kernels = channels * depth * height * width;
+
+  const int threads = 512;
+  const int blocks = DIVUP(num_kernels, threads);
+
+  keMatrixCol2Vol<<<blocks, threads>>>(num_kernels,
```
Comment (chengduoZH) [inline paddle/cuda/src/hl_cuda_matrix.cu]: Done

File: paddle/gserver/layers/Conv3DLayer.cpp [hunk 20]
```diff
@@ -0,0 +1,240 @@
...
+    const ConvConfig &conf = inputConfig.conv_conf();
+    M_.push_back(numFilters_ / conf.groups());
+    K_.push_back(filterPixels_[index] * filterChannels_[index]);
+    if (nullptr != weights_[index]->getW())
+      weights_[index]->getW()->reshape(weights_[index]->getW()->getWidth(),
+                                       weights_[index]->getW()->getHeight());
+    if (nullptr != weights_[index]->getWGrad())
+      weights_[index]->getWGrad()->reshape(
+          weights_[index]->getWGrad()->getWidth(),
+          weights_[index]->getWGrad()->getHeight());
```
Comment (chengduoZH) [inline paddle/gserver/layers/Conv3DLayer.cpp]: Done

File: paddle/gserver/layers/ConvBaseLayer.cpp (hunk 6 above)
Comment (chengduoZH) [inline paddle/gserver/layers/ConvBaseLayer.cpp]: Done

File: paddle/gserver/layers/DeConv3DLayer.cpp (hunk 11 above)
Comment (chengduoZH) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: Done. Sorry, I'm not familiar with ./paddle/math/MathUtils.h

File: paddle/gserver/layers/DeConv3DLayer.cpp (hunk 8 above)
Comment (chengduoZH) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: Done

File: paddle/gserver/layers/DeConv3DLayer.cpp (hunk 7 above)
Comment (chengduoZH) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: Done

File: paddle/gserver/layers/DeConv3DLayer.cpp (hunk 9 above)
Comment (chengduoZH) [inline paddle/gserver/layers/DeConv3DLayer.cpp]: Done

File: paddle/gserver/tests/test_LayerGrad.cpp (hunk 12 above)
Comment (chengduoZH) [inline paddle/gserver/tests/test_LayerGrad.cpp]: It has been modified

File: paddle/math/tests/test_matrixCompare.cpp (hunk 13 above)
Comment (chengduoZH) [inline paddle/math/tests/test_matrixCompare.cpp]: Done

File: paddle/math/tests/test_matrixCompare.cpp (hunk 14 above)
Comment (chengduoZH) [inline paddle/math/tests/test_matrixCompare.cpp]: Done

File: paddle/cuda/include/hl_matrix.h [hunk 21]
```diff
@@ -224,4 +224,80 @@ extern void hl_matrix_collect_shared_bias(real* B_d,
...
+ * @param[in]   filterD    depth of filter.
+ * @param[in]   filterH    height of filter.
+ * @param[in]   filterW    width of filter.
+ * @param[in]   strideD    stride in the depth.
+ * @param[in]   strideH    stride in the height.
+ * @param[in]   strideW    stride in the width.
+ * @param[in]   paddingD   padding in the depth.
+ * @param[in]   paddingH   padding in the height.
+ * @param[in]   paddingW   padding in the width.
+ * @param[out]   matDst     output matrix.
```
Comment (chengduoZH) [inline paddle/cuda/include/hl_matrix.h]: thanks, I'll fix it the next time I submit it.

File: paddle/gserver/layers/Conv3DLayer.cpp (hunk 5 above)
Comment (chengduoZH) [inline paddle/gserver/layers/Conv3DLayer.cpp]: As previously mentioned,  images in different batch may not be equal in training,  so These lines of code may be modified later.

File: python/paddle/trainer/config_parser.py [hunk 22]
```diff
@@ -1895,6 +2000,163 @@ class ConvLayer(ConvLayerBase):
...
+            self.config.num_filters = num_filters
+
+        use_gpu = int(g_command_config_args.get("use_gpu", 0))
+        parallel_nn = int(g_command_config_args.get("parallel_nn", 0))
+
+        # Automatically select cudnn_type for GPU and exconv for CPU
+        # if set type=conv, but still reserve the way user specify
+        # exconv or cudnn_conv manually.
+        if self.layer_type == "cudnn_conv3d":
+            config_assert(use_gpu, "cudnn_conv3d only support GPU")
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: 这里不对，`cudnn_conv3d` 目前还没有实现

File: python/paddle/trainer/config_parser.py [hunk 23]
```diff
@@ -1277,6 +1327,50 @@ def parse_conv(conv, input_layer_name, conv_conf, num_filters, trans=False):
...
+            get_img3d_size(input_layer_name, conv.channels)
+        conv_conf.output_x = cnn_output_size(
+            conv_conf.img_size, conv_conf.filter_size, conv_conf.padding,
+            conv_conf.stride, conv_conf.caffe_mode)
+        conv_conf.output_y = cnn_output_size(
+            conv_conf.img_size_y, conv_conf.filter_size_y, conv_conf.padding_y,
+            conv_conf.stride_y, conv_conf.caffe_mode)
+        conv_conf.output_z = cnn_output_size(
+            conv_conf.img_size_z, conv_conf.filter_size_z, conv_conf.padding_z,
+            conv_conf.stride_z, conv_conf.caffe_mode)
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]:
```
inputs = [get_img3d_size(input_layer_name, conv.channels)]
outputs =[conv_conf.output_x, conv_conf.output_y, conv_conf.output_z]
filters = [conv_conf.filter_size, conv_conf.filter_size_y, ...]
paddings = [conv_conf.padding, ..., ...]
strides = [...,...,...]

for out, in, filter, padding, stride in zip(outputs, inputs, filters, paddings, strides):
     out = cnn_output_size(in, filter, filter, filter, conv_conf.caffe_mode)

```
下同

File: python/paddle/trainer/config_parser.py [hunk 24]
```diff
@@ -1895,6 +2000,163 @@ class ConvLayer(ConvLayerBase):
...
 
 
+@config_layer('conv_3d')
+class Conv3DLayerBase(LayerBase):
+    def __init__(self,
+                 name,
+                 inputs=[],
+                 bias=True,
+                 num_filters=None,
+                 shared_biases=False,
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: `shared_biases=True`

File: python/paddle/trainer/config_parser.py [hunk 25]
```diff
@@ -1895,6 +2000,163 @@ class ConvLayer(ConvLayerBase):
...
+            self.config.num_filters = num_filters
+
+        use_gpu = int(g_command_config_args.get("use_gpu", 0))
+        parallel_nn = int(g_command_config_args.get("parallel_nn", 0))
+
+        # Automatically select cudnn_type for GPU and exconv for CPU
+        # if set type=conv, but still reserve the way user specify
+        # exconv or cudnn_conv manually.
+        if self.layer_type == "cudnn_deconv3d":
+            config_assert(use_gpu, "cudnn_conv3d only support GPU")
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: 和上面同样问题，还没实现cudnn_deconv3d

File: python/paddle/trainer/config_parser.py [hunk 26]
```diff
@@ -1895,6 +2000,163 @@ class ConvLayer(ConvLayerBase):
...
+        if shared_biases:
+            psize = self.config.num_filters
+        self.create_bias_parameter(bias, psize, [psize, 1])
+
+    def calc_parameter_size(self, conv_conf):
+        return self.config.num_filters * conv_conf.filter_channels \
+               * (conv_conf.filter_size * conv_conf.filter_size_y \
+                  * conv_conf.filter_size_z)
+
+    def set_layer_height_width(self, depth, height, width):
```
Comment (qingqing01) [inline python/paddle/trainer/config_parser.py]: `set_layer_height_width` 名字不合适

而且LayerBase里已经有 `set_layer_height_width` 和 `set_layer_depth`了吧。

File: python/paddle/trainer_config_helpers/layers.py [hunk 27]
```diff
@@ -6210,3 +6141,182 @@ def kmax_sequence_score_layer(input, name=None, beam_size=1):
...
+                     shared_biases=True,
+                     layer_attr=None,
+                     filter_size_y=None,
+                     stride_y=None,
+                     padding_y=None,
+                     filter_size_z=None,
+                     stride_z=None,
+                     padding_z=None,
+                     trans=False,
+                     layer_type=None):
```
Comment (qingqing01) [inline python/paddle/trainer_config_helpers/layers.py]: 感觉要么去掉 `_y`, `_z`的参数， `filter_size`，`stride`等接受collections.Sequence类型吧

**Pull request merged**